### PR TITLE
Replace existing prototype RNNT

### PR DESCRIPTION
**Pull request by carolineechen**
Summary: Replace the current torchaudio RNNT (using warp-transducer) to one without external library dependencies

Differential Revision: D28094572

Follow ups:
- check the status of the output in `cpu_transducer.h`
- improve `csrc` files (better `.h` `.cpp` organization, variable renaming)
- migrate input checks to C++

### CODE REVIEW

File: torchaudio/csrc/CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,16 @@ set(
...
 if(BUILD_TRANSDUCER)
-  list(APPEND LIBTORCHAUDIO_SOURCES transducer.cpp)
+  set(
+    TRANSDUCER_SOURCES
+    rnnt/cpu/compute_alphas.cpp
+    rnnt/cpu/compute_betas.cpp
+    rnnt/cpu/compute.cpp
+    rnnt/compute_alphas.cpp
+    rnnt/compute_betas.cpp
+    rnnt/compute.cpp
```
Comment (vincentqb) [inline torchaudio/csrc/CMakeLists.txt:21]: remind me the differences between the duplicate files?

File: torchaudio/csrc/rnnt/cpu/cpu_transducer.h [hunk 2]
```diff
@@ -36,15 +36,15 @@ status_t Compute(
   const int& D = options.numTargets_;
 
   { // compute denominators.
-    status_t status = LogSumExp2D<DTYPE, CAST_DTYPE>(
```
Comment (vincentqb) [inline torchaudio/csrc/rnnt/cpu/cpu_transducer.h]: let's mark as follow-up for later to make sure to check the status of the output, and whether any error followed

**Comment by vincentqb** [comment]
is [this](https://app.circleci.com/pipelines/github/pytorch/audio/5847/workflows/5451853e-65f3-4cce-b065-63994c7f8519/jobs/204182) failing on master too ?

**Review by vincentqb — APPROVED** [review]
Awesome! LGTM :) [The review was done mostly internally on the stack of diffs.]

**Comment by vincentqb** [comment]
> is [this](https://app.circleci.com/pipelines/github/pytorch/audio/5847/workflows/5451853e-65f3-4cce-b065-63994c7f8519/jobs/204182) failing on master too ?

yup

File: torchaudio/csrc/rnnt/cpu/cpu_kernels.h [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+    seqlogProbs.push_back(TensorView<const LogProbs<CAST_DTYPE>>(
+        {maxT, maxU},
+        reinterpret_cast<LogProbs<CAST_DTYPE>*>(
+            const_cast<CAST_DTYPE*>(logProbs)) +
+            b * maxT * maxU));
+    seq_betas.push_back(
+        TensorView<CAST_DTYPE>({maxT, maxU}, betas + b * maxT * maxU));
+  }
+
+  std::vector<CAST_DTYPE> scores(B << 1);
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/cpu_kernels.h:484]: This variable can be removed as it is never used.

File: torchaudio/csrc/rnnt/cpu/cpu_kernels.h [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+  //#pragma omp parallel for
+  for (int i = 0; i < B; ++i) { // use max 2 * B threads.
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/cpu_kernels.h:486]: The comment `// use max 2 * B threads.` is confusing. I think removing it
will improve readability.

File: torchaudio/csrc/rnnt/cpu/cpu_kernels.h [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+    seqlogProbs.push_back(TensorView<const LogProbs<CAST_DTYPE>>(
+        {maxT, maxU},
+        reinterpret_cast<LogProbs<CAST_DTYPE>*>(
+            const_cast<CAST_DTYPE*>(logProbs)) +
+            b * maxT * maxU));
+    seq_alphas.push_back(
+        TensorView<CAST_DTYPE>({maxT, maxU}, alphas + b * maxT * maxU));
+  }
+
+  std::vector<CAST_DTYPE> scores(B << 1);
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/cpu_kernels.h:447]: `scores` is never used.

File: torchaudio/csrc/rnnt/cpu/compute.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  c10::optional<torch::Tensor> gradients = c10::nullopt;
+  if (logits.requires_grad()) {
+    if (reuse_logits_for_grads) {
+      gradients = logits;
+    } else {
+      gradients = torch::zeros_like(logits);
+    }
+  }
+
+  torch::Tensor int_workspace = torch::empty(
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/compute.cpp:43]: For the CPU version, why do we need to allocate space for `int_workspace`?
Looks like the memory is allocated but is never used.

File: torchaudio/csrc/rnnt/cpu/compute_alphas.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      torch::TensorOptions().device(logits.device()).dtype(logits.dtype()));
+
+  torch::Tensor int_workspace = torch::empty(
+      IntWorkspace::ComputeSizeFromOptions(options),
+      torch::TensorOptions()
+          .device(logits.device())
+          .dtype(torch::ScalarType::Int));
+
+  torch::Tensor float_workspace = torch::empty(
+      DtypeWorkspace<float>::ComputeSizeFromOptions(options),
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/compute_alphas.cpp:40]: The allocated space for `beta` is wasteful here, I think.

File: torchaudio/csrc/rnnt/cpu/compute_betas.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      torch::TensorOptions().device(logits.device()).dtype(logits.dtype()));
+
+  torch::Tensor int_workspace = torch::empty(
+      IntWorkspace::ComputeSizeFromOptions(options),
+      torch::TensorOptions()
+          .device(logits.device())
+          .dtype(torch::ScalarType::Int));
+
+  torch::Tensor float_workspace = torch::empty(
+      DtypeWorkspace<float>::ComputeSizeFromOptions(options),
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/compute_betas.cpp:44]: Also, we don't need to allocate space for `alpha` here.

File: torchaudio/csrc/rnnt/cpu/cpu_kernels.h [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+  const int& blank = options.blank_;
+  const CAST_DTYPE clamp = options.clamp_;
+
+  CAST_DTYPE cost = -beta({0, 0});
+
+  // Note - below gradient is different from numpy_transducer, since we
+  // compute log_softmax more efficiently within the loss, to save memory The
+  // details of the below implementation / equations can be found in Sec 3.2
+  // (function merging) in below paper:
+  // https://www.microsoft.com/en-us/research/uploads/prod/2019/10/RNNT.pdf
```
Comment (csukuangfj) [inline torchaudio/csrc/rnnt/cpu/cpu_kernels.h:321]: Is there any plan to implement another trick mentioned in this paper to reduce memory consumption?

Comment (carolineechen) [reply torchaudio/csrc/rnnt/cpu/cpu_kernels.h:321]: @csukuangfj thanks for the code comments -- I'll take a look at them

there currently isn't a plan to implement something specifically from that paper, although we are interested in further improving RNNT loss
